### PR TITLE
feat: add release action

### LINKED ISSUE #112: Provide major (and minor) tags
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Provides major tag `v2` and possible `v3` in the future which tag to the latest appropriate commit.
(Side note: I saw you provides `v1` which tags to the same commit as `v1.0.0`)

### Motivation

As it is stated in GitHub Docs <https://docs.github.com/en/actions/creating-actions/releasing-and-maintaining-actions#example-developer-process>

> We recommend creating releases using semantically versioned tags – for example, `v1.1.3` – and keeping major (`v1`) and minor (`v1.1`) tags current to the latest appropriate commit. For more information, see "[About custom actions](https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-release-management-for-actions)" and "[About semantic versioning](https://docs.npmjs.com/about-semantic-versioning).

### Example

_No response_

### PR DESCRIPTION
**Pull request by nuragic**
Close #112.

### CODE REVIEW

**Review by Eomm — APPROVED** [review]

**Comment by simoneb** [comment]
perfect. please also link the issue so that it's closed when we merge this PR

**Comment by nuragic** [comment]
@simoneb I've linked the issue in the commit description

**Review by simoneb — APPROVED** [review]

**Review by simoneb — CHANGES_REQUESTED** [review]
we can't use this action yet, more details to follow

**Comment by simoneb** [comment]
I'm blocking this PR because I just recalled that there's something that the release automation action doesn't do for us and we need to decide how to handle it, or this won't do what we expect.

This action has a [build npm script](https://github.com/fastify/github-action-merge-dependabot/blob/main/package.json#L7). This script needs to be executed in order to generate the dist folder which is what is actually being executed when the action runs.

The release automation action does npm install (this thing has a problem in itself because it runs on whatever Node.js version the action runs on, NOT what we would like it to run on, but that's another issue), but it runs no other scripts. Without running the build script, the published released will not contain the changes it would need to contain.

I see 2 options here:

- we enhance the release action to run additional scripts (action-side approach)
- we configure this repo to run the build script on a npm lifecycle script, e.g. prepare or postinstall or whatever makes sense (user-side approach)

**Comment by simoneb** [comment]
Opened an issue about this in https://github.com/nearform/optic-release-automation-action/issues/27

**Comment by Eomm** [comment]
> we enhance the release action to run additional scripts (action-side approach)

Just to share that, by default, the GitHub nodejs action template provides you this workflow to check the `dist` build:

https://github.com/Eomm/why-don-t-you-tweet/blob/main/.github/workflows/check-dist.yml

This is a reminder to build the source code actually.

**Comment by nuragic** [comment]
Great, thanks for the feedback! Given the above, I think it would make sense to go with:
> we enhance the release action to run additional scripts (action-side approach)

Just a question aside, why we compile everything via `ncc`?

**Comment by simoneb** [comment]
> Just a question aside, why we compile everything via `ncc`?

The action needs to be self contained, so either we commit node_modules or we bundle it. This second approach is also what's recommended in https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action#commit-tag-and-push-your-action-to-github.

In fact, we should probably we doing the same for the release action! I'll create an issue.

**Comment by nuragic** [comment]
Oh thanks for the link!

> Checking in your node_modules directory can cause problems. As an alternative, you can use a tool called @vercel/ncc to compile your code and modules into one file used for distribution.

👍

**Comment by simoneb** [comment]
There's one more thing that we have to do before we go ahead with this PR. We first need to release a new version capable of automerging semver-like. So we should keep doing manual releases until https://github.com/fastify/github-action-merge-dependabot/issues/124 is released.

**Comment by nuragic** [comment]
@simoneb now that #124 is released, should we unblock this PR?

**Comment by simoneb** [comment]
@nuragic no but thanks for bringing it up because the dependency was not explicit. We also depend on https://github.com/nearform/optic-release-automation-action/issues/27

**Comment by simoneb** [comment]
@nuragic @Eomm is there a chance that in fact we're already good here?

See https://github.com/fastify/github-action-merge-dependabot/blob/main/.husky/pre-commit

**Comment by Eomm** [comment]
I don't think we have done here: the local hooks can be skipped and a malicious user may manipulate the `dist/` build (who is reading that file)

When we will release this PR https://github.com/nearform/optic-release-automation-action/pull/30 we are ready to go here.
The actual `optic-release-automation-action` version runs the `actions/checkout@v2.4.0` step every time overwriting whatever build the user made before.. so those PR's changes are necessary to let the GHA environment build a secure `dist/` payload

File: .github/workflows/release.yml
Comment (Eomm) [inline .github/workflows/release.yml:16]:
```suggestion
    steps:
      - name: Setting action build runtime
        uses: actions/setup-node@v2
        with:
          node-version: 16
```

File: .github/workflows/release.yml
Comment (Eomm) [inline .github/workflows/release.yml:17]:
```suggestion
      - uses: nearform/optic-release-automation-action@v2
```

File: .github/workflows/release.yml
Comment (Eomm) [inline .github/workflows/release.yml:21]:
```suggestion
          sync-semver-tags: true
          build-command: |
            echo 'installing husky hooks..'
            npm ci

```

**Review by Eomm — CHANGES_REQUESTED** [review]

**Review by Eomm — APPROVED** [review]

**Review by simoneb — APPROVED** [review]

**Pull request merged**
